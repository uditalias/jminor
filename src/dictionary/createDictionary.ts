import createDefaultKeyGenerator from "../generators/defaultKeyGenerator";
import { isArray, isObject, reverse, clone } from "../utils";
import IKeyGenerator from "../generators/IKeyGenerator";

export class Dictionary {
    private _dict: { [index: string]: string };
    private _rdict: { [index: string]: string };

    constructor(private _keyGen: IKeyGenerator) {
        this._dict = {};
        this._rdict = {};
    }

    public replaceKeyGenerator(keyGenerator: IKeyGenerator): Dictionary {
        this._keyGen = keyGenerator;
        this._dict = {};
        this._rdict = {};

        return this;
    }

    public fromJSON(data: any): Dictionary {
        this._generateKeys(data);
        this._rdict = reverse(this._dict);

        return this;
    }

    public extendWith(data: any): Dictionary {
        this._generateKeys(data);
        this._rdict = reverse(this._dict);

        return this;
    }

    public export(): { [index: string]: string } {
        return clone(this._dict);
    }

    public import(rawDictionary: { [index: string]: string }): Dictionary {
        this._dict = clone(rawDictionary);
        this._rdict = reverse(this._dict);

        this._setGeneratorLastKey();

        return this;
    }

    public ktoc(key: string): string {
        return this._dict[key] || key;
    }

    public ctok(ckey: string): string {
        return this._rdict[ckey] || ckey;
    }

    private _setGeneratorLastKey() {
        const ckeys = Object.keys(this._rdict);

        ckeys.sort(this._keyGen.sortKeys);

        this._keyGen.setLastKey(ckeys[ckeys.length - 1]);
    }

    private _generateKeys(obj: any) {
        for (const key in obj) {
            if (!this._dict[key]) {
                this._dict[key] = this._keyGen.next();
            }

            if (isArray(obj[key])) {
                obj[key].forEach((x) => this._generateKeys(x));
            } else if (isObject(obj[key])) {
                this._generateKeys(obj[key]);
            }
        }
    }
}


export default function createDictionary(
    createKeyGenerator: () => IKeyGenerator = createDefaultKeyGenerator
): Dictionary {
    return new Dictionary(createKeyGenerator());
}