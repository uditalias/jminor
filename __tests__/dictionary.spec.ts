'use strict';

import createDictionary from "../lib/dictionary/createDictionary"
import createDefaultKeyGenerator from "../lib/generators/defaultKeyGenerator";
import createNumericKeyGenerator from "../lib/generators/numericKeyGenerator";

describe('defaultKeyGenerator', () => {

    it('should create dictionary', () => {

        const dictionary = createDictionary();

        expect(dictionary).toBeDefined();
    });

    it('should import raw dictionary data', () => {

        const dictionary = createDictionary(createNumericKeyGenerator).import({
            first: "1",
            second: "2"
        });

        expect(dictionary.ctok("1")).toBe("first");
        expect(dictionary.ctok("2")).toBe("second");
    });

    it('should export raw dictionary', () => {

        const dictionary = createDictionary().fromJSON({
            first: 1,
            second: [{
                third: 3
            }],
            fourth: {
                fifth: {
                    sixth: 6,
                    seventh: 7
                },
                eighth: 8
            },
            ninth: 9
        });

        expect(dictionary.export()).toMatchSnapshot();
    });

    it('should extend dictionary with more data', () => {
        const dictionary = createDictionary().import({
            first: "a",
            second: "b"
        });

        dictionary.extendWith({
            second: [{
                third: 3,
                fourth: 4
            }],
            fifth: 5
        });

        expect(dictionary.export()).toMatchSnapshot();
    });

    it('should get ckey for key', () => {
        const dictionary = createDictionary().import({
            first: "a",
            second: "b"
        });

        expect(dictionary.ktoc("first")).toBe("a");
        expect(dictionary.ktoc("third")).toBe("third");
    });

    it('should get key for ckey', () => {
        const dictionary = createDictionary().import({
            first: "a",
            second: "b"
        });

        expect(dictionary.ctok("a")).toBe("first");
        expect(dictionary.ctok("third")).toBe("third");
    });

    it('should set key generator', () => {
        const dictionary = createDictionary();

        dictionary.setKeyGenerator(createDefaultKeyGenerator())
    });

});