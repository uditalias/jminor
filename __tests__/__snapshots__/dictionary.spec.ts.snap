// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultKeyGenerator should export raw dictionary 1`] = `
Object {
  "eighth": "h",
  "fifth": "e",
  "first": "a",
  "fourth": "d",
  "ninth": "i",
  "second": "b",
  "seventh": "g",
  "sixth": "f",
  "third": "c",
}
`;

exports[`defaultKeyGenerator should extend dictionary with more data 1`] = `
Object {
  "fifth": "e",
  "first": "a",
  "fourth": "d",
  "second": "b",
  "third": "c",
}
`;
